### PR TITLE
Issue #15 Use status file to get the IP and remove to look for leasefile

### PR DESCRIPTION
**Pull request by praveenkumar**
We don't need to have now lookupIPFromLeasesFile since all the major
distro have libvirt > 2.0.0 and that was for the 1.2.6

Also we are adding 120sec wait time in 40 loop in 3 sec interval then
default one.

### CODE REVIEW

File: libvirt.go [hunk 1]
```diff
@@ -553,23 +540,45 @@ func (d *Driver) getIPByMacFromSettings(mac string) (string, error) {
...
+
+func (d *Driver) lookupIP() (string, error) {
+	conn, err := d.getConn()
+	if err != nil {
+		return "", fmt.Errorf("%v : getting connection and domain", err)
+	}
+
+	defer conn.Close()
+
+	libVersion, err := conn.GetLibVersion()
```
Comment (cfergeau) [inline libvirt.go]: We don't want to get/print libvirt version every time we call lookupIP(), it's not going to change over the lifetime of the application anyway.

File: libvirt.go [hunk 2]
```diff
@@ -553,23 +540,45 @@ func (d *Driver) getIPByMacFromSettings(mac string) (string, error) {
...
+	log.Debugf("Libvirt version: %s", libVersion)
+
+	mac, err := d.getMAC()
+	if err != nil {
+		return "", fmt.Errorf("%v : Not able to get mac address", err)
 	}
 
-	return ip, err
+	// TODO: for everything > 1002006, there is direct support in the libvirt-go for handling this
+	return d.getIPByMacFromSettings(mac)
```
Comment (cfergeau) [inline libvirt.go:557]: libvirt-go is still parsing some internal libvirt files to achieve this. I believe the libvirt C library offers an API to directly get the IP rather than having to do it ourselves in libvirt.go

File: libvirt.go [hunk 3]
```diff
@@ -316,17 +316,28 @@ func (d *Driver) Start() error {
 	// They wont start immediately
 	time.Sleep(5 * time.Second)
 
-	for i := 0; i < 90; i++ {
-		time.Sleep(time.Second)
-		ip, _ := d.GetIP()
+	for i := 0; i <= 40; i++ {
```
Comment (cfergeau) [inline libvirt.go]: I would split all these changes from the getIPByMacFromSettings change as they are not related

File: libvirt.go (hunk 2 above)
Comment (praveenkumar) [reply libvirt.go:557]: @cfergeau I don't have time right now to check in the C library implementation part and check if libvirt-go wrapper have same functionality. I can put a TODO here and we can circle back to it if that makes sense?

File: libvirt.go (hunk 3 above)
Comment (praveenkumar) [reply libvirt.go]: Yes I will do that.

**Comment by praveenkumar** [comment]
@cfergeau I broken the commit messages, have a look again.

File: libvirt.go [hunk 4]
```diff
@@ -558,15 +534,7 @@ func (d *Driver) GetIP() (string, error) {
...
-	 * TODO - Figure out what version of libvirt changed behavior and
-	 *        be smarter about selecting which algorithm to use
-	 */
-	ip, err := d.getIPByMACFromLeaseFile(mac)
-	if ip == "" {
-		ip, err = d.getIPByMacFromSettings(mac)
-	}
-
+	ip, err := d.getIPByMacFromSettings(mac)
 	return ip, err
```
Comment (cfergeau) [inline libvirt.go]: these 2 lines can be condensed in `return d.getIPByMacFromSettings(mac)`

File: libvirt.go [hunk 5]
```diff
@@ -530,6 +530,13 @@ func (d *Driver) getIPByMacFromSettings(mac string) (string, error) {
 
 func (d *Driver) GetIP() (string, error) {
 	log.Debugf("GetIP called for %s", d.MachineName)
+	s, err := d.GetState()
+	if err != nil {
+		return "", fmt.Errorf("%v : machine in unknown state", err)
+	}
+	if s != state.Running {
+		return "", errors.New("host is not running")
+	}
```
Comment (cfergeau) [inline libvirt.go:551]: what kind of false results do we get if the VM is not running?

File: libvirt.go [hunk 6]
```diff
@@ -318,7 +318,10 @@ func (d *Driver) Start() error {
 
 	for i := 0; i < 90; i++ {
 		time.Sleep(time.Second)
-		ip, _ := d.GetIP()
+		ip, err := d.GetIP()
+		if err != nil {
+			return fmt.Errorf("%v: getting ip during machine start", err)
+		}
 		if ip != "" {
 			// Add a second to let things settle
```
Comment (cfergeau) [inline libvirt.go]: I think this loop is ignoring errors on purpose, and is busy looping while it waits for the VM to start. Do you have a specific error in mind when we could return early because we know we are never recover from that error?

File: libvirt.go [hunk 7]
```diff
@@ -316,20 +316,29 @@ func (d *Driver) Start() error {
 	// They wont start immediately
```
Comment (cfergeau) [inline libvirt.go:316]: Not clear from the commit log why it's desirable to retry for 120s rather than 90s?

File: libvirt.go [hunk 8]
```diff
@@ -316,20 +316,29 @@ func (d *Driver) Start() error {
...
 		if err != nil {
 			return fmt.Errorf("%v: getting ip during machine start", err)
 		}
+
+		if ip == "" {
+			log.Debugf("Waiting for machine to come up %d/%d", i, 40)
+			time.Sleep(3 * time.Second)
+			continue
+		}
+
```
Comment (cfergeau) [inline libvirt.go:330]: I would move this block to the very end of the loop where we currently have `log.Debugf("Waiting for the VM to come up... %d", i)` this should make the code easier to follow

File: libvirt.go [hunk 9]
```diff
@@ -316,20 +316,29 @@ func (d *Driver) Start() error {
...
 		if ip != "" {
-			// Add a second to let things settle
-			time.Sleep(time.Second)
-			return nil
+			log.Infof("Found IP for machine: %s", ip)
+			d.IPAddress = ip
```
Comment (cfergeau) [inline libvirt.go:333]: ah ok this also adds some caching of the IP address in the Driver structure. Are you sure we don't need to keep the 'sleep' though?

File: libvirt.go (hunk 5 above)
Comment (praveenkumar) [reply libvirt.go:551]: It is not about false result but exit as soon as possible we don't even have to go to next step if the VM is not running.

File: libvirt.go (hunk 8 above)
Comment (praveenkumar) [reply libvirt.go:330]: @cfergeau this block is already correct place.

```golang
if ip == "" {
			log.Debugf("Waiting for machine to come up %d/%d", i, 40)
			time.Sleep(3 * time.Second)
			continue
 }
```

File: libvirt.go (hunk 9 above)
Comment (praveenkumar) [reply libvirt.go:333]: @cfergeau once we get the IP we don't need any sleep and return it to driver.

Comment (cfergeau) [reply libvirt.go:333]: I don't know, there was a comment implying it's desirable to wait a little bit more after getting the IP. I'm worried by not waiting anymore, we'll be introducing some races ?

File: libvirt.go (hunk 8 above)
Comment (cfergeau) [reply libvirt.go:330]: if you move it to the end of the loop, you can get rid of the `if ip == ""` and of the `continue` without (I think) any changes of when this code is run. This makes the code easier to follow.

File: libvirt.go (hunk 6 above)
Comment (cfergeau) [reply libvirt.go]: Still the same comment, I believe the whole point of the loop is to wait until everything is up. If you are reasonably sure that we should be looping while we get `"", nil` from `GetIP()`, but that if we get `"", err` from `GetIP()`, then this is never going to be a transient error (ie the error won't go away once the boot process has progressed enough), then please explain this in the commit log.

File: libvirt.go (hunk 9 above)
Comment (praveenkumar) [reply libvirt.go:333]: Alright if that make you bit uncomfortable, I will add that sleep :)

File: libvirt.go (hunk 8 above)
Comment (praveenkumar) [reply libvirt.go:330]: @cfergeau Now I got it what you meant, will update it.

File: libvirt.go [hunk 10]
```diff
@@ -316,20 +316,28 @@ func (d *Driver) Start() error {
...
 			// Add a second to let things settle
 			time.Sleep(time.Second)
-			return nil
+			log.Infof("Found IP for machine: %s", ip)
+			d.IPAddress = ip
+			break
 		}
-		log.Debugf("Waiting for the VM to come up... %d", i)
+
+		log.Debugf("Waiting for machine to come up %d/%d", i, 40)
```
Comment (cfergeau) [inline libvirt.go]: This is starting to be very small details, but it would be better to define a const for the number of loops we are going to run, as if we want to change the number of iterations, we are bound to forget this one ;)

File: libvirt.go [hunk 11]
```diff
@@ -316,20 +316,28 @@ func (d *Driver) Start() error {
...
+		time.Sleep(3 * time.Second)
+		continue
```
Comment (cfergeau) [inline libvirt.go]: The continue can be removed.

File: libvirt.go (hunk 6 above)
Comment (cfergeau) [reply libvirt.go]: No input on this one?

**Review by cfergeau — APPROVED** [review]
I still would like more details in the commit log for `Handle error for GetIP instead ignoring it` (what we discussed, why this is not going to introduce a regression). Apart from this, it looks good to me.
